rfcs/039-table-data-quality-checks.md

```diff
@@ -0,0 +1,61 @@
+- Feature Name: table_data_quality_checks
+- Start Date: 2021-07-16
+- RFC PR: [amundsen-io/rfcs#XX](https://github.com/amundsen-io/rfcs/pull/XX)
+# Table Data Quality Checks
+
+## Summary
+This RFC will define a generic interface for accessing data quality checks in Amundsen. If available, we aim to expose
```

Comment:
I think it would be beneficial to take a slightly more comprehensive approach where we define the underlying data models that Amundsen can store internally for data quality checks, even if those are not immediately implemented. The thought here being that if we take the approach where we only define an interface for accessing data quality checks (presumably from an external system) it could possibly constrain the approach (or cause significant refactor) that we eventually take for storing similar data within Amundsen. 

I'd like to see us look at this from a slightly broader perspective to ensure that the design will be highly reusable, especially considering that there are several major players in the DQ space.

Reply:
I see your point about the value of a deeper integration with reusable and generic models.

One caveat about DQ checks is that live data is highly valuable compared to stale data that was ingested in a databuilder job some hours ago. For this reason we think accessing an external API that can provide live data would be preferable for this integration. 

We could create our own model for DQ checks in Amundsen and then pull updates from a service periodically, or on demand, but I'm not sure how much value that would bring for the current use case.

Reply:
I think another approach here could be to provide an endpoint that those quality services could call and send latest DQ check info. As stated in another comment, I think there are a lot of tools that can be used to run DQ checks but don't provide integration in the form of API to query those results and this approach limits us to only tools that would have it.

I also disagree with the point that databuilder is not a good fit here. A lof of DQ unit tests might be run with scheduling service like Airflow and it could be downstream task to ingest the data to Amundsen after running DQ test. The delay of 1-2 minutes is not something I would consider to be an issue.

In our case we run unit tests with Great Expectations using Airflow and if Amundsen enables DQ checks surfacing I would just register a callback ingesting this live to Atlas using databuidler job.

Reply:
Too add to this, Amundsen is well positioned to be able to propagate data quality downstream using the lineage information available. While I think that it is fine to enrich the data displayed in the UI using an external API to retrieve some more in-depth information, Amundsen should be able to take the information that it knows about failed DQ tests and do something with that information on downstream assets (e.g. place an alert on affected dashboards, etc.).

Reply:
DQ normally requires low sla which requires a push model while Amundsen / databuilder still has a pull model approach. How could the downstream consumer depends on Amundsen metadata in this case?

Reply:
I think that databuilder can really be considered push or pull based on the context in which it is executed. For example, for data quality, a company may have several teams that each use airflow to orchestrate the DAGs that do something like:

- Task 1: load a table
- Task 2: run DQ checks
- Task 3: push DQ results to Amundsen

Push / pull semantics aside, Amundsen would receive the data as soon as it is available. The specific cadence for which this DQ process runs actually does not matter, whether it is hourly, daily, weekly, etc. As soon as Amundsen receives an update that the DQ checks have failed it should take action to propagate that information downstream.

In regards to the SLA, from both the user perspective, as well as Amundsen's perspective, the state of the data quality is the last known value that was generated, regardless of whether or not the actual underlying data has changed. Therefore, the actual SLA for DQ is not important, what is important is that DQ events are made available to Amundsen as soon as they are created.

Reply:
These are wonderful questions, but as I've tried to point out below they are far outside the scope of what we were proposing - which is on us for not being as precise as we should have been. Please take a look at my thoughts here: https://github.com/amundsen-io/rfcs/pull/41#issuecomment-892064322

We'll iterate on the RFC to make its intention more clear. 

In the meantime, our proposed work will take no direction one way or the other on the above.